snowdrop-core/src/Snowdrop/Core/Validator/Basic.hs

```diff
@@ -1,39 +1,58 @@
 module Snowdrop.Core.Validator.Basic
-       ( StructuralValidationException (..)
+       ( valid
+       , validateIff
+       , validateAll
+
+       , StructuralValidationException (..)
        , structuralPreValidator
 
-       , inputsExist
-       , inputsSigned
-       , exampleStateTxValidator
-       , redundantIdsPreValidator
-       , RedundantIdException (..)
+       -- , inputsExist
+       -- , inputsSigned
+       -- , exampleStateTxValidator
        ) where
 
+import           Control.Monad.Except (MonadError (..))
 import           Universum
 
-import qualified Data.List.NonEmpty as NE
 import qualified Data.Map.Strict as M
-import qualified Data.Set as S
 import qualified Data.Text.Buildable
+import           Data.Vinyl.Core (Rec (..))
 import           Formatting (bprint, build, (%))
 
-import           Snowdrop.Core.ChangeSet.Type (changeSetToList, csNew, csRemove)
-import           Snowdrop.Core.ERoComp.Helpers (query, validateAll, validateIff)
-import           Snowdrop.Core.Prefix (IdSumPrefixed (..), Prefix (..))
-import           Snowdrop.Core.Transaction (TxProof, txBody, txProof)
-import           Snowdrop.Core.Validator.Types (PreValidator (..), Validator, mkValidator)
+import           Snowdrop.Core.ChangeSet (HChangeSet, csNew, csRemove)
+import           Snowdrop.Core.ERoComp (ERoComp, QueryERo, query)
+import           Snowdrop.Core.Transaction (TxComponents, txBody)
+import           Snowdrop.Core.Validator.Types (PreValidator (..))
 import           Snowdrop.Util
 
+---------------------------
+-- Helpers for validation
+---------------------------
+
+valid :: (Monoid a, Monad m) => m a
+valid = pure mempty
+
+validateIff :: forall e e1 m a . (Monoid a, MonadError e m, HasReview e e1) => e1 -> Bool -> m a
+validateIff e1 = bool (throwLocalError e1) valid
+
+validateAll
+    :: (Foldable f, MonadError e m, Container (f a))
+    => (Element (f a) -> e)
+    -> (Element (f a) -> Bool)
+    -> f a
+    -> m ()
+validateAll ex p ls = maybe (pure ()) (throwError . ex) (find (not . p) ls)
+
 ---------------------------
 -- Structural validator
 ---------------------------
 
-data StructuralValidationException id
-    = DpRemoveDoesntExist id
-    | DpAddExist id
-    deriving (Show)
+data StructuralValidationException
+    = forall id . (Show id, Buildable id) => DpRemoveDoesntExist id
```

Comment:
What about holding `Text` there instead of `id`?

Reply:
meh :( let's change it later...